File: Python/890 Find and Replace Pattern.py
# Find and Replace Pattern
class Solution:
    def findAndReplacePattern(self, words, pattern):
        """
        :type words: List[str]
        :type pattern: str
        :rtype: List[str]
        """
        ans = []
        for word in words:
            dic1 = {}
            dic2 = {}
            flag = True
            for i in range(len(word)):
                dic1[word[i]] = dic1.get(word[i], pattern[i])
                dic2[pattern[i]] = dic2.get(pattern[i],word[i])
                if dic1[word[i]] != pattern[i] or dic2[pattern[i]]!=word[i]:
                    flag = False
                    break
            if flag:
                ans.append(word)
                
        return ans

File: C++/890 Find and Replace Pattern.cpp
class Solution {
public:
    vector<string> findAndReplacePattern(vector<string>& words, string pattern) {
        vector<string> ans;
        for (string word :words) {
            bool flag = true;
            if (word.length() != pattern.length()) continue;
            unordered_map<char, char> m1;
            unordered_map<char, char> m2;
            for (int i = 0; i < word.length(); i++) {
                char w = word[i];
                char p = pattern[i];
                if (!m1.count(w)) m1.insert({w,p});
                if (!m2.count(p)) m2.insert({p,w});
                if (m1[w]!=p || m2[p]!=w) flag = false;
            }
            if (flag) ans.push_back(word);
        }
        return ans;
    }
};